docs/rules/data-testid.md

```diff
@@ -0,0 +1,46 @@
+# Enforces consistent naming for the data-testid attribute (data-testid)
+
+Ensure `data-testid` values match a provided regex. This rule is un-opinionated, and requires configuration.
+
+## Rule Details
+
+> Assuming the rule has been configured with the following regex: `^TestId(\_\_[A-Z]*)?$`
+
+Examples of **incorrect** code for this rule:
+
+```js
+const foo = props => <div data-testid="my-test-id">...</div>;
+const foo = props => <div data-testid="myTestId">...</div>;
+const foo = props => <div data-testid="TestIdEXAMPLE">...</div>;
+```
+
+Examples of **correct** code for this rule:
+
+```js
+const foo = props => <div data-testid="TestId__EXAMPLE">...</div>;
+
+const bar = props => <div data-testid="TestId">...</div>;
+
+const baz = props => <div>...</div>;
+```
+
+## Options
+
+| Option        | Details                                                                                                                                                                                                                                                           | Example                                       |
+| ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------- |
+| testIdPattern | A regex used to validate the format of the `data-testid` value. `{componentName}` can optionally be used as a placeholder and will be substituted with the name of the file OR the name of the files parent directory in the case when the fileName is `index.js` | `'^{componentName}(\_\_([A-Z]+[a-z]_?)+)_\$'` |
```

Comment:
Shouldn't be `fileName` instead of `componentName`? 🤔

Reply:
Indeed. What would `componentName` be if the component is `MyComponent` but the file is called `my-component.js`?

Reply:
Yep agreed I think that makes more sense here. 

Ideally we could figure out the component name via the AST but that's pretty non-trivial and expensive. Internally we enforce file structure and component organization to make sure each component is contained within it's own file, so this implicit filename -> component name assumption works for us there - but that is definitely not a universal assumption so I will change this, thanks for pointing it out!

Also, this was also originally inspired from the BEM stylelint plugin, which uses similar logic to infer componentName in scss files (https://github.com/postcss/postcss-bem-linter#define-components-and-utilities-implicitly-based-on-their-filename).